vision/odlc_contour_filtering.py

```diff
@@ -0,0 +1,498 @@
+    box: bbox.BoundingBox = min_common_bounding_box([contour, approx])
+
+    # generates masks (single channel binary images/matricies) for both shapes with white being
+    # any point that is in the shape and black being everywhere else
+    # the dimensions of the masks will be the same (both equal to the dimensions of the previously
+    # found box)
+    contour_mask: consts.Mask = generate_mask(contour, box)
+    approx_mask: consts.Mask = generate_mask(approx, box)
+
+    # makes a new mask where white is any point that was in one shape but not the other
+    # aka a logical_xor between the two matricies of booleans
+    non_overlap_mask: consts.Mask = np.logical_xor(contour_mask, approx_mask)
+
+    # converts the boolean image to a single channel 8-bit image (still binarized with 0 and 255)
+    non_overlap_img: consts.ScImage = non_overlap_mask.astype(UInt8)
+    # detects all of the new shapes made by the xor operation
+    non_overlap_cnts: tuple[consts.Contour]
+    non_overlap_cnts, _ = cv2.findContours(non_overlap_img, cv2.RETR_TREE, cv2.CHAIN_APPROX_SIMPLE)
+
+    # sums the area of all of the non-overlapping portions of the shapes
+    non_overlap_area_sum = 0
+    for cnt in non_overlap_cnts:
+        non_overlap_area_sum += cv2.contourArea(cnt)
+
+    contour_area = cv2.contourArea(contour)
+    if (
+        len(non_overlap_cnts) == 0
+        or non_overlap_area_sum == 0
+        or 1 - test_percent_diff < contour_area / non_overlap_area_sum < 1 + test_percent_diff
+    ):
+        return True
+    return False
+
+
+def test_circleness(img: consts.ScImage) -> bool:
+    """
+    This function will test the cropped area around a given contour to see if it is a circle or if
+    it is not a circle, it uses the cv2.HoughCircles() function to check for a circle
+
+    Parameters
+    ----------
+    img : consts.ScImage
+        Img should be grayscale cropped box around contour and PROCESSED ALREADY
+        as in the image is binary with the contour filled in as white
+        NOTE: sc_image_type denotes a single channel image
+
+    Returns
+    -------
+    is_circular : bool
+        Returns true if circle of appropriate size is found (to reduce chance of false positives)
+    """
+    padding: int = int(img.shape[0] * 0.05)
+    modded: consts.ScImage = cv2.copyMakeBorder(
+        img, padding, padding, padding, padding, cv2.BORDER_CONSTANT, None, 0
+    )
+    # 5 chosen arbitrarily as the kernel size and sigmaX params for the blur function
+    # (kernel size and sigmaX are both parameters of cv2.GaussianBlur())
+    # the number should not be very important because the input image is a binarized image
+    modded = cv2.GaussianBlur(img, (5, 5), 5)
+    # NDArray[Shape["1, *, 3"], Float32] is return type of cv2.HoughCircles()
+    # format is [[[center_x_1, center_y_1, radius_1], [center_x_2, center_y_2, radius_2], ...]]
+    circles: NDArray[Shape["1, *, 3"], Float32] | None = cv2.HoughCircles(
```

Comment:
First comment here about hough circles return type isn't necessary since you have the type hint